iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1043 @@
-                    case TransportState.Opening:
-                        _fatalException = ExceptionDispatchInfo.Capture(exception);
-                        _connectCompletion.TrySetException(exception);
-                        _subscribeCompletionSource.TrySetException(exception);
-                        break;
-
-                    case TransportState.Open:
-                    case TransportState.Subscribing:
-                        _fatalException = ExceptionDispatchInfo.Capture(exception);
-                        _subscribeCompletionSource.TrySetException(exception);
-                        OnTransportDisconnected();
-                        break;
-
-                    case TransportState.Receiving:
-                        _fatalException = ExceptionDispatchInfo.Capture(exception);
-                        _disconnectAwaitersCancellationSource.Cancel();
-                        OnTransportDisconnected();
-                        break;
+                // Note the request as "in progress" before actually sending it so that no matter how quickly the service
+                // responds, this layer can correlate the request.
+                _inProgressGetTwinRequests.Add(requestId);
+                MqttClientPublishResult result = await _mqttClient.PublishAsync(mqttMessage, cancellationToken);
 
-                    default:
-                        Debug.Fail($"Unknown transport state: {previousState}");
-                        throw new InvalidOperationException();
+                if (result.ReasonCode != MqttClientPublishReasonCode.Success)
+                {
+                    throw new IotHubClientException($"Failed to publish the mqtt packet for getting this client's twin with reason code {result.ReasonCode}", true);
                 }
-
-                await _closeRetryPolicy.RunWithRetryAsync(CleanUpImplAsync).ConfigureAwait(true);
             }
-            catch (Exception ex) when (!Fx.IsFatal(ex))
+            catch (Exception)
             {
-                if (Logging.IsEnabled)
-                    Logging.Error(this, ex.ToString(), nameof(OnError));
+                // Since the request failed due to a network level issue or was rejected by the service, stop tracking
+                // this request id.
+                _inProgressGetTwinRequests.Remove(requestId);
+                throw;
             }
-            finally
+
+            Logging.Info(this, $"Sent twin get request with request id {requestId}. Now waiting for the service response.");
+            while (!getTwinResponses.ContainsKey(requestId))
             {
-                if (Logging.IsEnabled)
-                    Logging.Exit(this, exception, nameof(OnError));
+                // May need to wait multiple times. This semaphore is released each time a get twin
+                // request gets a response, but it may not be in response to this particular get twin request.
+                _getTwinSemaphore.Wait(cancellationToken);
```

Comment:
This makes me slightly nervous. Are there multiple listeners? Are other twin responses getting lost?

Reply:
There could be multiple listeners, yes. A user could call getTwin from multiple threads on the same device client instance

Reply:
I get that involving semaphores gets scary quickly, but I can't think of another way to make this scenario work. We send a message that gets a basic ack, but we need to then wait for the service to send a message to the client that contains the twin. We could make this a callback, but that seems like an implementation detail that would frustrate users.

Reply:
The concurrent dictionary has its own semaphores. It is thread safe.

Also, this loop could be very tight. Aren't you worried about over taxing the process checking so fast?

Reply:
There is no blocking call to get an item from a concurrent dictionary, though. The semaphore here isn't for protecting the resource saving these responses. It is for waiting for the service's response to the request this client's invocation of getTwin sent.

Reply:
If you know of a ```concurrentDictionary.BlockingGet("myKey");``` then that would be perfect here. I haven't found such an implementation in .NET though

Reply:
I'll take a look at how we handled this in AMQP since we should have the same problem over there

Reply:
No way to be notified by the concurrent dictionary for an element.

What if the call to wait never returns?

Reply:
Then we time out with the cancellation token and throw operation cancelled exception up to the user. That would only happen if the service never sent the twin response message, though

Reply:
We use TaskCompletionSource in our AMQP layer as the way to wait on another thread to receive a message, so I'll adopt that pattern here as well.